corehq/apps/userreports/reports/builder/forms.py

```diff
@@ -846,14 +816,64 @@ def _get_user_name_pseudo_property():
             data_types=["string"],
         )
 
+
+class ApplicationCaseDataSourceHelper(CaseDataSourceHelper):
+    def __init__(self, domain, app, source_type, source_id):
+        self.domain = domain
+        self.app = app
+        self.source_type = source_type
+
+        # case type or form ID
+        self.source_id = source_id
+        assert source_type == 'case'
+        prop_map = get_case_properties(
+            self.app, [self.source_id], defaults=list(DEFAULT_CASE_PROPERTY_DATATYPES),
+            include_parent_properties=True,
+        )
+        self.case_properties = sorted(set(prop_map[self.source_id]) | {'closed', 'closed_on'})
+
+    def _get_data_source_build_information(self):
+        return DataSourceBuildInformation(
+            source_id=self.source_id,
+            app_id=self.app._id,
+            app_version=self.app.version,
+        )
+
     @property
     @memoized
     def data_source_name(self):
         today = datetime.datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S")
         return "{} (v{}) {}".format(self.source_id, self.app.version, today)
 
 
-def get_data_source_interface(domain, app, source_type, source_id):
+class RegistryCaseDataSourceHelper(CaseDataSourceHelper):
+    def __init__(self, domain, registry_slug, source_type, source_id):
+        assert source_type == 'case'
+
+        self.domain = domain
+        self.registry_slug = registry_slug
+        self.source_type = source_type
+        self.source_id = source_id
+
+        owning_domain = DataRegistry.objects.get(slug=self.registry_slug).domain
+        prop_map = get_data_dict_props_by_case_type(owning_domain)
+        self.case_properties = sorted(set(prop_map[self.source_id]) | {'closed', 'closed_on'})
+
+    def _get_data_source_build_information(self):
+        return DataSourceBuildInformation(
+            source_id=self.source_id,
```

Comment:
we should add the registry slug to this meta class so that we can use it when editing the report (see `ConfigureNewReportBase._bootstrap`)